[Translation] Add icu message support for the lokalise provider

DanielBadura opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 5.4 <!-- see below -->
| Bug fix?      | yes
| New feature?  | no <!-- please update src/**/CHANGELOG.md files -->
| Deprecations? | no <!-- please update UPGRADE-*.md and src/**/CHANGELOG.md files -->
| Tickets       | - <!-- prefix each issue number with "Fix #", no need to create an issue if none exist, explain below instead -->
| License       | MIT
| Doc PR        | - <!-- required for new features -->

This PR would add the support for the ICU message format when using the lokalise provider. I saw that the unit tests are written as it would support the ICU message format so therefore the PR aims the 5.4 branch as a bugfix.

The API from lokalise needs a flag for the format to be downloaded, otherwise it return the file in the "classic" format with `%` as the delimiter for the placeholders. Also the file was always saved without the suffix. For this i looked up how the ICU support was implemented in symfony and tried to adapt this approach.

I think i would need some assitance how to write tests with and without the intl extension. Right now, some tests fail since the intl extension in enabled and i did not adjust them yet.

DanielBadura committed:
Add icu message support for the lokalise provider

DanielBadura committed:
fix the codestyle

DanielBadura commented:
I looked up how the extraction command is handling the ICU thing, and there, if i undestand it correctly, all message will extracted as ICU. The only exception is when the translation key is already in a domain which is not ICU. So maybe my approach here is wrong to begin with. So i would really appreciate some help or advices to the right direction! :)

welcoMattic commented on src/Symfony/Component/Translation/Bridge/Lokalise/LokaliseProvider.php:
```diff
@@ -45,6 +46,7 @@ public function __construct(HttpClientInterface $client, LoaderInterface $loader
         $this->logger = $logger;
         $this->defaultLocale = $defaultLocale;
         $this->endpoint = $endpoint;
+        $this->icuEnabled = class_exists(\MessageFormatter::class);
```
The existence of `MessageFormatter` class does not mean that the user has to use the ICU format. I think this is a false assertion, and one that may cause problems for some people.

AFAIK, we never find any universal right way to determine if we have to write translation in ICU format or not. The only cases where we can guess the format is when there are existing translation file, so we check the extension. Otherwise, we must let the user decide to use or not ICU format.

In this case, we can check on `--intl-icu` option from TranslationPullCommand (https://github.com/symfony/symfony/blob/6.1/src/Symfony/Component/Translation/Command/TranslationPullCommand.php#L98) and try to pass it to the provider (most of them doesn't need it, but it's ok for me to forward the information to the providers).

cc @nicolas-grekas

DanielBadura replied:
That some useful information. Sounds right for me to use the option, didnt even seen it before. But i think adding the information to the provider would be best at initiation but this is not possible if the re-use the option from the command. So maybe it would also be a solution to able to configure it with the other provider configs?

welcoMattic replied:
It can be configured for each provider in `translation.yaml`, yes. But be careful, if a provider is configured to use intl-icu format, and `translation:pull` is called without `--intl-icu` option, we have to set the right filename.

DanielBadura replied:
So what i think is we have 2 options here.

1. Use the `--intl-icu` option from the `TranslationPullCommand` and forward this information to the Providers. 
a) This could happen via an extra argument for the `write` and `read` methods
b) via a configuration method and mutate the state of the service, this one i really dislike

2. Create a new configuration option for the translator component with which the user can then decide if the icu format should be used or not. This could be used on other places too, where symfony right now tries to guess which format is appropiate. This config could be injected into the providers and they can then do whatever they need to enable the icu feature.

Did i miss another option?

The other question i have: Is it right to assume that this is a bug or is it more a feature? Im somewhat not sure at this topic :)

welcoMattic replied:
I think we must limit to translation providers for now. So, the 2. option seems the right one, but by create the new configuration for each provider:

```yaml

framework:
    default_locale: en
    translator:
        default_path: '%kernel.project_dir%/translations'
        fallbacks:
            - en
        providers:
            lokalise:
                dsn: '%env(CROWDIN_DSN)%'
                intl_icu: true # or false
```

For the question about bug or feature, in case of doubt it's always a feature.

DanielBadura replied:
For my perspective, since icu support should already been in there, some tests try to check it, it's a bug. But i wanted the pov from the maintainer e.g. you. Then i would move the base branch if this is not considered a bug.

Why do we need to limit this to the provider? I mean, the config could be on "component level" but at first only used for the providers.

DanielBadura replied:
I tried to add the configuration for providers. But i added the flag first only for the Lokalise Factory / Provider. Maybe you could have a look again? Thanks!

DanielBadura committed:
remove icu detection instead user can configure it now for the provider

DanielBadura committed:
update tests and add new ones

DanielBadura committed:
cs fix

DanielBadura committed:
fix integration test cases caused by adjusting ProviderTestCase by adding the intlIcuEnabled parameter

DanielBadura commented:
I am still having problems even with these changes and try to figure it out with the Lokalise support. The `%LANG_ISO%` placeholder is not added to the filenames for example.

DanielBadura commented:
So after a longer discussion with the Lokalise support i got to this version of the LokaliseProvider:

<details>
<summary>The Diff</summary>
<p>

```diff
diff --git a/src/Symfony/Component/Translation/Bridge/Lokalise/LokaliseProvider.php b/src/Symfony/Component/Translation/Bridge/Lokalise/LokaliseProvider.php
index b86d3924ff..b431d263bf 100644
--- a/src/Symfony/Component/Translation/Bridge/Lokalise/LokaliseProvider.php
+++ b/src/Symfony/Component/Translation/Bridge/Lokalise/LokaliseProvider.php
@@ -14,6 +14,7 @@ namespace Symfony\Component\Translation\Bridge\Lokalise;
 use Psr\Log\LoggerInterface;
 use Symfony\Component\Translation\Exception\ProviderException;
 use Symfony\Component\Translation\Loader\LoaderInterface;
+use Symfony\Component\Translation\MessageCatalogue;
 use Symfony\Component\Translation\MessageCatalogueInterface;
 use Symfony\Component\Translation\Provider\ProviderInterface;
 use Symfony\Component\Translation\TranslatorBag;
@@ -37,14 +38,16 @@ final class LokaliseProvider implements ProviderInterface
     private $logger;
     private $defaultLocale;
     private $endpoint;
+    private $intlIcuEnabled;

-    public function __construct(HttpClientInterface $client, LoaderInterface $loader, LoggerInterface $logger, string $defaultLocale, string $endpoint)
+    public function __construct(HttpClientInterface $client, LoaderInterface $loader, LoggerInterface $logger, string $defaultLocale, string $endpoint, bool $intlIcuEnabled = false)
     {
         $this->client = $client;
         $this->loader = $loader;
         $this->logger = $logger;
         $this->defaultLocale = $defaultLocale;
         $this->endpoint = $endpoint;
+        $this->intlIcuEnabled = $intlIcuEnabled;
     }

     public function __toString(): string
@@ -102,7 +105,14 @@ final class LokaliseProvider implements ProviderInterface

         foreach ($translations as $locale => $files) {
             foreach ($files as $filename => $content) {
-                $translatorBag->addCatalogue($this->loader->load($content['content'], $locale, str_replace('.xliff', '', $filename)));
+                $domain = str_replace('.' . $locale . '.xliff', '', $filename);
+
+                if ($this->intlIcuEnabled) {
+                    $domain .= MessageCatalogue::INTL_DOMAIN_SUFFIX;
+                }
+
+                $catalogue = $this->loader->load($content['content'], $locale, $domain);
+                $translatorBag->addCatalogue($catalogue);
             }
         }

@@ -149,6 +159,8 @@ final class LokaliseProvider implements ProviderInterface
                 'filter_langs' => array_values($locales),
                 'filter_filenames' => array_map([$this, 'getLokaliseFilenameFromDomain'], $domains),
                 'export_empty_as' => 'skip',
+                'placeholder_format' => $this->intlIcuEnabled ? 'icu' : 'symfony',
+                'include_tags' => ['symfony'],
             ],
         ]);

@@ -183,6 +195,7 @@ final class LokaliseProvider implements ProviderInterface
                     'android' => null,
                     'other' => null,
                 ],
+                'tags' => ['symfony'],
             ];
         }

@@ -352,6 +365,6 @@ final class LokaliseProvider implements ProviderInterface

     private function getLokaliseFilenameFromDomain(string $domain): string
     {
-        return sprintf('%s.xliff', $domain);
+        return sprintf('%s.%s.xliff', $domain, '%LANG_ISO%');
     }
 }
```
</p>
</details>

In this diff are 3 changes, 2 of them which are imho important, the last one less important but nice to have.

1. The most important the ICU Change proposed in this PR waiting for comments if this is a bugfix and if the configuration suggestion is good to go.
2. The `%LANG_ISO%` placeholder Lokalise uses for replacement of the language suffix. We need to add this placeholder ourself. Lokalise cannot detect it via API right now, that's why i'm also replacing here the locale: `$domain = str_replace('.' . $locale . '.xliff', '', $filename);`. This should also be a bugfix imho.
3. Adding the possibility to add a tag for the translation keys. With this developers can better filter the translations via Lokalise features. We are needing this since we are using Lokalise also with our frontend and there we are not needing the backend / symfony translations. But i think this should also be optional and not hardcoded like it is in the diff. That should be a new feature and is the less important one imho.

What do you think about these points? I would happy to create these into issues / PR's.

fabpot commented:
@welcoMattic Can you have a look at the latest comment from @DanielBadura?

welcoMattic commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php:
```diff
@@ -856,6 +856,7 @@ private function addTranslatorSection(ArrayNodeDefinition $rootNode, callable $e
                                 ->fixXmlConfig('locale')
                                 ->children()
                                     ->scalarNode('dsn')->end()
+                                    ->booleanNode('enable_intl_icu')->defaultFalse()->end()
```
Can you rename it to `intl_icu_enabled` to keep consistency with the variable name in the code?

welcoMattic commented:
@DanielBadura 

> So after a longer discussion with the Lokalise support i got to this version of the LokaliseProvider:
> The Diff
> 
> In this diff are 3 changes, 2 of them which are imho important, the last one less important but nice to have.
> 
> 1. The most important the ICU Change proposed in this PR waiting for comments if this is a bugfix and if the configuration suggestion is good to go.
> 
> 2. The `%LANG_ISO%` placeholder Lokalise uses for replacement of the language suffix. We need to add this placeholder ourself. Lokalise cannot detect it via API right now, that's why i'm also replacing here the locale: `$domain = str_replace('.' . $locale . '.xliff', '', $filename);`. This should also be a bugfix imho.
> 

For the first 2 points, I'm ok with it, you did great work, thank you. Just a small comment about parameter naming.

> 3. Adding the possibility to add a tag for the translation keys. With this developers can better filter the translations via Lokalise features. We are needing this since we are using Lokalise also with our frontend and there we are not needing the backend / symfony translations. But i think this should also be optional and not hardcoded like it is in the diff. That should be a new feature and is the less important one imho.

> What do you think about these points? I would happy to create these into issues / PR's.


About this, I have been wondering for some time if a particular configuration per provider is something to add. And your comment goes in this direction: for example, we could allow users to define one or more tags to add to their Symfony translations, but without forcing it.

We should consider it as a feature to be done in another PR.

DanielBadura committed:
rename configuration to match it with the variablename in the implementation

welcoMattic commented:
@DanielBadura could you have a look to failing checks? Otherwise, I this PR is ready to merge for me

DanielBadura commented:
Yes i will try to look at it the next days @welcoMattic :)

DanielBadura committed:
provider are default with intl-icu disabled

DanielBadura commented:
So i rebased again and added a default value for the provider creation for the tests. That resolved most of the failing test. Now only my new tests are failing :D. But I'm really not sure why I went through it mutltiple times now and i dont understand why the test is failing.. Maybe someone can also have a look? 😅

welcoMattic commented:
@DanielBadura I'll try to take a look on it this week

welcoMattic commented on src/Symfony/Component/Translation/Bridge/Lokalise/Tests/LokaliseProviderTest.php:
```diff
@@ -601,4 +603,119 @@ public function getResponsesForManyLocalesAndManyDomains(): \Generator
+        $loader = $this->getLoader();
+        $loader->expects($this->once())
+            ->method('load')
+            ->willReturn((new XliffFileLoader())->load($responseContent, $locale, $domain));
```
```suggestion
            ->willReturn((new XliffFileLoader())->load($responseContent, $locale, $domain.MessageCatalogueInterface::INTL_DOMAIN_SUFFIX));
```

welcoMattic replied:
This will fix the tests :wink:

DanielBadura replied:
Thanks! :D

DanielBadura committed:
Fix test due to missing domain suffix

Co-authored-by: Mathieu Santostefano <msantostefano@protonmail.com>

DanielBadura commented:
code style checks from fabbot seems unrelated 🤔

nicolas-grekas commented:
Sorry to come late in the review, but can't we use the intl-icu suffix in the domain to decide for the correct format?

DanielBadura commented:
Maybe this is possible for sending the files based on that. But if we have no files local and want to read the files from the provider we need to pass the option to get it in the right format.

nicolas-grekas commented:
Can't we set some metadata on the provider side when pushing to know the format?

DanielBadura commented:
Im not sure how this can help us since we need to know the format before the query the provider. The format is part of the inital request to retrieve the data. If i understand correctly what you try to achieve is that lokalise is giving us the data in a specific format if we tell them so in some kind of metadata when uploading. AFAIK that is not a feature lokalise e.g. is offering. Maybe with their automation feature the user can somehow achieve that but Im not sure about that. Or did I misunderstood you?

nicolas-grekas commented:
You've got me right. Is lokalize the only provider that allow fetching the messages in various formats? I feel like the other providers don't deal with formats and give back strings as is. Am I right?
What about putting a dummy message as a marker that ICU was used in the source? Would that work as a kind of metadata?
I'm thinking loud and I don't know the topic well, sorry if that doesn't make sense?

DanielBadura commented:
I can not answer this question about the other providers since i only worked with `Lokalise` so far. But i looked briefly into the documentation for `Loco` and there it seems they also offer this. See here the export API we are using in the LocoProvider: https://localise.biz/api/docs/export/exportall and the param doc https://localise.biz/help/developers/printf 

Side note: The namings for all the providers can be confusing.
* `Lokalise` (the provider we are talking about for this PR): https://lokalise.com/
* `Loco` aka. `localise`: https://localise.biz/
* `Localize`: https://localizejs.com

So i assumed you meant `Lokalise` with `lokalize`.

nicolas-grekas commented:
I'm going to delegate to @welcoMattic at this point :grimacing:

welcoMattic commented:
@DanielBadura yes, the Provider names are confusing, but we can not do anything about that except to be rigorous when mentioning them.

About the main problem, which I understand is telling Lokalise if we want to download translation under ICU format or "symfony" format, IIRC Lokalise is the only provider (for now) which lets us define the format at download by exposing a parameter in the query, so let's use it. 

About @nicolas-grekas comment:
> Sorry to come late in the review, but can't we use the intl-icu suffix in the domain to decide for the correct format?

No we can't, because at the provider level, we only know the domain itself, not the variant (with or without `+intl-icu` suffix). It on purpose, to avoid translations duplication on providers side when user switch from the old format to ICU.

So, for me the only way to do, is the one has been done here: expose a config option for Lokalise provider, to explicitly ask for ICU format or "symfony" format. 

Now, one question remains: should we set the config option at `icu` or `symfony` by default?

nicolas-grekas commented:
icu if extensio_loadede('intl')?

welcoMattic commented:
@DanielBadura about Loco (localise.biz), they also expose a parameter to force icu format at export (see https://localise.biz/api/docs/export/exportall). Are you ok to open another PR to add support for icu format on Loco Provider, after this one?

DanielBadura commented:
Well since I'm not working with Loco or generally with translations right now, I couldnt really test it. 😅

welcoMattic commented:
> Well since I'm not working with Loco or generally with translations right now, I can not really test it. 😅 

Ok, no problem, I will take it

TwanVermeulen commented on src/Symfony/Component/Translation/Bridge/Lokalise/LokaliseProvider.php:
```diff
@@ -149,6 +156,7 @@ private function exportFiles(array $locales, array $domains): array
                 'filter_langs' => array_values($locales),
                 'filter_filenames' => array_map([$this, 'getLokaliseFilenameFromDomain'], $domains),
                 'export_empty_as' => 'skip',
+                'placeholder_format' => $this->intlIcuEnabled ? 'icu' : 'symfony',
```
A different parameter is required to ensure that the plurals are also in the expected ICU format. The param is called `plural_format` which should contain the same value as  placeholder_format. See https://developers.lokalise.com/reference/download-files for all params. It would be great if that can be included. Without this option users might get invalid ICU translation files

TwanVermeulen commented on src/Symfony/Component/Translation/Bridge/Lokalise/LokaliseProvider.php:
```diff
@@ -149,6 +156,7 @@ private function exportFiles(array $locales, array $domains): array
                 'filter_langs' => array_values($locales),
                 'filter_filenames' => array_map([$this, 'getLokaliseFilenameFromDomain'], $domains),
                 'export_empty_as' => 'skip',
+                'placeholder_format' => $this->intlIcuEnabled ? 'icu' : 'symfony',
```
```suggestion
                'placeholder_format' => $this->intlIcuEnabled ? 'icu' : 'symfony',
                'plural_format' => $this->intlIcuEnabled ? 'icu' : 'symfony',
```

A different parameter is required to ensure that the plurals are also in the expected ICU format. The param is called `plural_format` which should contain the same value as  placeholder_format. See https://developers.lokalise.com/reference/download-files for all params. It would be great if that can be included. Without this option users might get invalid ICU translation files

TwanVermeulen replied:
Without this change, you will still get plurals in the old format from Lokalise, which the Symfony Translator is not expecting in ICU files

DanielBadura replied:
This would be really weird behaviour IMHO, but better safe then sorry i guess :D

Could you verify this?

welcoMattic replied:
It's weird, but I guess @TwanVermeulen is right. According to the documentation https://developers.lokalise.com/reference/api-plurals-and-placeholders#plural-format-descriptions I guess we have to pass both parameters.

DanielBadura replied:
I can add it, but as i said - cant really verify for now as I'm not working anymore on the project which needed this patch. So is this verified that this is needed and working?

TwanVermeulen reviewed (commented):
I've added a suggestion that will fix a possible issue for the plurals

DanielBadura commented:
So what could be the next steps to finish this PR? IIRC we did not conclude on how the configuration should be done.

welcoMattic commented:
@DanielBadura sorry for the delay. To finish this PR I think we need to be sure about the plural parameter value.

About the configuration, I think we are ok. We need to enable the ICU format per Provider. As I mentioned earlier, I will take care of Loco Provider once this PR will be merged.

TwanVermeulen commented:
I'm not sure if it helps, but I did verify this myself + I got the confirmation from the Lokalise support desk. If you want to receive the plurals in ICU format, you need to specify this via plural_format configuration.

nicolas-grekas commented:
I'd like to have an answer to my previous comment:
>But if the provider is in control, we don't care about the exact format on the app's side, do we? If not, we should fetch the most capable format, which is ICU. Why would the app want to decide what it works with when it can delegate to the provider?

WDYT? I'm not happy with adding a config option...

DanielBadura commented:
IMHO the app is in control in which format it is using the translations. I'm not sure how we want to delegate this to the provider to decide 🤔.

welcoMattic commented:
> I'd like to have an answer to my previous comment:
> 
> > But if the provider is in control, we don't care about the exact format on the app's side, do we? If not, we should fetch the most capable format, which is ICU. Why would the app want to decide what it works with when it can delegate to the provider?
> 
> WDYT? I'm not happy with adding a config option...

Some apps are still  working with the old plural format, not ICU. For those apps, we must keep an option to not fetch translations in ICU format, otherwise I guess the fetched translations will be in a weird format (old plural format treated as ICU).

Did I miss something?

DanielBadura commented:
How can we move forward here?

welcoMattic commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php:
```diff
@@ -857,6 +857,7 @@ private function addTranslatorSection(ArrayNodeDefinition $rootNode, callable $e
                                 ->fixXmlConfig('locale')
                                 ->children()
                                     ->scalarNode('dsn')->end()
+                                    ->booleanNode('intl_icu_enabled')->defaultFalse()->end()
```
```suggestion
                                    ->booleanNode('intl_icu')->defaultFalse()->end()
```

I think it's better to name it short.

welcoMattic commented:
> Some apps are still working with the old plural format, not ICU. For those apps, we must keep an option to not fetch translations in ICU format, otherwise I guess the fetched translations will be in a weird format (old plural format treated as ICU).
> Did I miss something?

I'm still convinced that we must leave the control of ICU or not to the app. For example, an app with the special `%count%` parameter in a `trans` call will fail with ICU formatted translation, IIRC the difference between both formats.

The configuration option seems to be our only solution here. It's totally acceptable for me.

@nicolas-grekas @DanielBadura WYDT?

To move forward, this branch needs to be rebased. @DanielBadura, are the functional tests still relevant? (I mean, if you test this branch with real API calls, Lokalise responds as expected?)

DanielBadura commented:
> I'm still convinced that we must leave the control of ICU or not to the app. For example, an app with the special %count% parameter in a trans call will fail with ICU formatted translation, IIRC the difference between both formats.

> The configuration option seems to be our only solution here. It's totally acceptable for me.

I think the same, hence why i proposed the configuration option. 

> To move forward, this branch needs to be rebased. @DanielBadura, are the functional tests still relevant? (I mean, if you test this branch with real API calls, Lokalise responds as expected?)

I could rebase this branch. About the testing part: I'm no longer working on the project which had this implementation, so i cant really test it right now with real API calls to Lokalise.
